File: sdk/python/cds/sources/finance.py
"""
SignedData CDS — Finance Brazil Ingestors
Sources:
    BCB SGS API: https://api.bcb.gov.br/dados/serie/bcdata.sgs.{code}/dados/ultimos/{n}
    Brapi:       https://brapi.dev/api/quote/{tickers}
    Copom:       https://www.bcb.gov.br/api/servico/sitebcb/reunioes-copom/listar
"""
from __future__ import annotations

import hashlib
from datetime import UTC, datetime
from typing import Any

import httpx

from cds.ingestor import BaseIngestor
from cds.schema import CDSEvent, ContextMeta, SourceMeta
from cds.sources.finance_models import (
    CDIRate,
    CopomDecision,
    FinanceContentTypes,
    FXRate,
    IGPMIndex,
    IPCAIndex,
    SELICRate,
    StockQuote,
)
from cds.vocab import CDSSources

BCB_SGS_BASE = "https://api.bcb.gov.br/dados/serie/bcdata.sgs.{code}/dados/ultimos/{n}"
BRAPI_BASE = "https://brapi.dev/api"
COPOM_URL = "https://www.bcb.gov.br/api/servico/sitebcb/reunioes-copom/listar"

# BCB SGS series codes
SGS_SELIC = 11
SGS_CDI = 4391
SGS_IPCA_MONTHLY = 433
SGS_IPCA_12M = 432
SGS_IGPM = 189
SGS_USD_BRL_BUY = 1
SGS_USD_BRL_SELL = 2
SGS_EUR_BRL = 21619


def _bcb_url(series_code: int, last_n: int = 1) -> str:
    return BCB_SGS_BASE.format(code=series_code, n=last_n)


def _parse_bcb_date(raw: str) -> str:
    """Convert BCB date 'DD/MM/YYYY' → 'YYYY-MM-DD'."""
    parts = raw.split("/")
    if len(parts) == 3:
        return f"{parts[2]}-{parts[1]}-{parts[0]}"
    return raw


def _daily_rate(annual_pct: float) -> float:
    """Convert annual rate (%) to daily rate using 252 business days."""
    return ((1 + annual_pct / 100) ** (1 / 252) - 1) * 100


def _fingerprint(data: bytes) -> str:
    return "sha256:" + hashlib.sha256(data).hexdigest()


# ── BCB Rates Ingestor ────────────────────────────────────


class BCBRatesIngestor(BaseIngestor):
    content_type = FinanceContentTypes.SELIC

    def __init__(self, signer: Any, last_n: int = 1) -> None:
        super().__init__(signer)
        self.last_n = last_n

    async def fetch(self) -> list[CDSEvent]:
        events: list[CDSEvent] = []
        async with httpx.AsyncClient(timeout=15) as client:
            # SELIC
            resp = await client.get(_bcb_url(SGS_SELIC, self.last_n))
            resp.raise_for_status()
            fp = _fingerprint(resp.content)
            for item in resp.json():
                date_iso = _parse_bcb_date(item["data"])
                annual = float(item["valor"])
                rate = SELICRate(
                    date=date_iso,
                    rate_annual=annual,
                    rate_daily=round(_daily_rate(annual), 6),
                    effective_date=date_iso,
                )
                events.append(CDSEvent(
                    content_type=FinanceContentTypes.SELIC,
                    source=SourceMeta(id=CDSSources.BCB_API, fingerprint=fp),
                    occurred_at=datetime.fromisoformat(f"{date_iso}T18:30:00-03:00"),
                    lang="pt-BR",
                    payload=rate.model_dump(mode="json"),
                    event_context=ContextMeta(
                        summary=f"SELIC: {rate.rate_annual}% a.a. ({rate.date})",
                        model="rule-based-v1",
                    ),
                ))

            # CDI
            resp = await client.get(_bcb_url(SGS_CDI, self.last_n))
            resp.raise_for_status()
            fp = _fingerprint(resp.content)
            for item in resp.json():
                date_iso = _parse_bcb_date(item["data"])
                annual = float(item["valor"])
                cdi = CDIRate(
                    date=date_iso,
                    rate_annual=annual,
                    rate_daily=round(_daily_rate(annual), 6),
                )
                events.append(CDSEvent(
                    content_type=FinanceContentTypes.CDI,
                    source=SourceMeta(id=CDSSources.BCB_API, fingerprint=fp),
                    occurred_at=datetime.fromisoformat(f"{date_iso}T18:30:00-03:00"),
                    lang="pt-BR",
                    payload=cdi.model_dump(mode="json"),
                    event_context=ContextMeta(
                        summary=f"CDI: {cdi.rate_annual}% a.a. ({cdi.date})",
                        model="rule-based-v1",
                    ),
                ))

        return events


# ── BCB Indices Ingestor ──────────────────────────────────


class BCBIndicesIngestor(BaseIngestor):
    content_type = FinanceContentTypes.IPCA

    def __init__(self, signer: Any, last_n: int = 1) -> None:
        super().__init__(signer)
        self.last_n = last_n

    async def fetch(self) -> list[CDSEvent]:
        events: list[CDSEvent] = []
        async with httpx.AsyncClient(timeout=15) as client:
            # IPCA monthly
            resp_m = await client.get(_bcb_url(SGS_IPCA_MONTHLY, self.last_n))
            resp_m.raise_for_status()
            # IPCA 12-month
            resp_12 = await client.get(_bcb_url(SGS_IPCA_12M, self.last_n))
            resp_12.raise_for_status()
            fp = _fingerprint(resp_m.content + resp_12.content)

            monthly_data = resp_m.json()
            acc_12m_data = resp_12.json()

            for m_item, a_item in zip(monthly_data, acc_12m_data):
                date_iso = _parse_bcb_date(m_item["data"])
                ipca = IPCAIndex(
                    date=date_iso,
                    monthly_pct=float(m_item["valor"]),
                    accumulated_12m=float(a_item["valor"]),
                    accumulated_year=0.0,  # requires additional calculation
                )
                events.append(CDSEvent(
                    content_type=FinanceContentTypes.IPCA,
                    source=SourceMeta(id=CDSSources.BCB_API, fingerprint=fp),
                    occurred_at=datetime.fromisoformat(f"{date_iso}T12:00:00-03:00"),
                    lang="pt-BR",
                    payload=ipca.model_dump(mode="json"),
                    event_context=ContextMeta(
                        summary=(
                            f"IPCA {date_iso}: {ipca.monthly_pct}% no mês, "
                            f"{ipca.accumulated_12m}% em 12 meses"
                        ),
                        model="rule-based-v1",
                    ),
                ))

            # IGP-M
            resp = await client.get(_bcb_url(SGS_IGPM, self.last_n))
            resp.raise_for_status()
            fp = _fingerprint(resp.content)
            for item in resp.json():
                date_iso = _parse_bcb_date(item["data"])
                igpm = IGPMIndex(
                    date=date_iso,
                    monthly_pct=float(item["valor"]),
                )
                events.append(CDSEvent(
                    content_type=FinanceContentTypes.IGPM,
                    source=SourceMeta(id=CDSSources.BCB_API, fingerprint=fp),
                    occurred_at=datetime.fromisoformat(f"{date_iso}T12:00:00-03:00"),
                    lang="pt-BR",
                    payload=igpm.model_dump(mode="json"),
                    event_context=ContextMeta(
                        summary=f"IGP-M {date_iso}: {igpm.monthly_pct}% no mês",
                        model="rule-based-v1",
                    ),
                ))

        return events


# ── BCB FX Ingestor ───────────────────────────────────────


class BCBFXIngestor(BaseIngestor):
    content_type = FinanceContentTypes.USD_BRL

    def __init__(self, signer: Any, last_n: int = 1) -> None:
        super().__init__(signer)
        self.last_n = last_n

    async def fetch(self) -> list[CDSEvent]:
        events: list[CDSEvent] = []
        async with httpx.AsyncClient(timeout=15) as client:
            # USD/BRL buy + sell
            resp_buy = await client.get(_bcb_url(SGS_USD_BRL_BUY, self.last_n))
            resp_buy.raise_for_status()
            resp_sell = await client.get(_bcb_url(SGS_USD_BRL_SELL, self.last_n))
            resp_sell.raise_for_status()
            fp = _fingerprint(resp_buy.content + resp_sell.content)

            for buy_item, sell_item in zip(resp_buy.json(), resp_sell.json()):
                date_iso = _parse_bcb_date(buy_item["data"])
                buy_val = float(buy_item["valor"])
                sell_val = float(sell_item["valor"])
                fx = FXRate(
                    date=date_iso,
                    buy=buy_val,
                    sell=sell_val,
                    mid=round((buy_val + sell_val) / 2, 4),
                    currency_from="USD",
                    currency_to="BRL",
                )
                events.append(CDSEvent(
                    content_type=FinanceContentTypes.USD_BRL,
                    source=SourceMeta(id=CDSSources.BCB_API, fingerprint=fp),
                    occurred_at=datetime.fromisoformat(f"{date_iso}T18:00:00-03:00"),
                    lang="pt-BR",
                    payload=fx.model_dump(mode="json"),
                    event_context=ContextMeta(
                        summary=(
                            f"USD/BRL PTAX {date_iso}: "
                            f"compra {fx.buy:.4f}, venda {fx.sell:.4f}"
                        ),
                        model="rule-based-v1",
                    ),
                ))

            # EUR/BRL
            resp = await client.get(_bcb_url(SGS_EUR_BRL, self.last_n))
            resp.raise_for_status()
            fp = _fingerprint(resp.content)
            for item in resp.json():
                date_iso = _parse_bcb_date(item["data"])
                val = float(item["valor"])
                fx = FXRate(
                    date=date_iso,
                    buy=val,
                    sell=val,
                    mid=val,
                    currency_from="EUR",
                    currency_to="BRL",
                )
                events.append(CDSEvent(
                    content_type=FinanceContentTypes.EUR_BRL,
                    source=SourceMeta(id=CDSSources.BCB_API, fingerprint=fp),
                    occurred_at=datetime.fromisoformat(f"{date_iso}T18:00:00-03:00"),
                    lang="pt-BR",
                    payload=fx.model_dump(mode="json"),
                    event_context=ContextMeta(
                        summary=f"EUR/BRL PTAX {date_iso}: {fx.mid:.4f}",
                        model="rule-based-v1",
                    ),
                ))

        return events


# ── Brapi Quotes Ingestor ─────────────────────────────────


class BrapiQuotesIngestor(BaseIngestor):
    content_type = FinanceContentTypes.STOCK

    def __init__(self, signer: Any, tickers: list[str] | None = None) -> None:
        super().__init__(signer)
        self.tickers = tickers or ["PETR4", "VALE3", "ITUB4", "BBDC4", "ABEV3"]

    async def fetch(self) -> list[CDSEvent]:
        events: list[CDSEvent] = []
        ticker_str = ",".join(self.tickers)
        async with httpx.AsyncClient(timeout=15) as client:
            resp = await client.get(f"{BRAPI_BASE}/quote/{ticker_str}")
            resp.raise_for_status()
            fp = _fingerprint(resp.content)
            data = resp.json()

            for item in data.get("results", []):
                quote = StockQuote(
                    ticker=item.get("symbol", ""),
                    short_name=item.get("shortName", ""),
                    long_name=item.get("longName", ""),
                    currency=item.get("currency", "BRL"),
                    market_price=float(item.get("regularMarketPrice", 0)),
                    change=float(item.get("regularMarketChange", 0)),
                    change_pct=float(item.get("regularMarketChangePercent", 0)),
                    previous_close=float(item.get("regularMarketPreviousClose", 0)),
                    open=item.get("regularMarketOpen"),
                    day_high=float(item.get("regularMarketDayHigh", 0)),
                    day_low=float(item.get("regularMarketDayLow", 0)),
                    volume=int(item.get("regularMarketVolume", 0)),
                    market_cap=item.get("marketCap"),
                    exchange=item.get("fullExchangeName", "SAO"),
                    market_state=item.get("marketState", "CLOSED"),
                    timestamp=item.get("regularMarketTime", datetime.now(UTC).isoformat()),
                )

                # Determine content type: FII tickers end with "11" and have specific patterns
                ct = FinanceContentTypes.STOCK
                if quote.ticker.endswith("11") and len(quote.ticker) == 6:
                    ct = FinanceContentTypes.FII

                occurred = datetime.now(UTC)
                try:
                    occurred = datetime.fromisoformat(quote.timestamp)
                except (ValueError, TypeError):
                    pass

                events.append(CDSEvent(
                    content_type=ct,
                    source=SourceMeta(id=CDSSources.BRAPI, fingerprint=fp),
                    occurred_at=occurred,
                    lang="pt-BR",
                    payload=quote.model_dump(mode="json"),
                    event_context=ContextMeta(
                        summary=(
                            f"{quote.ticker}: R$ {quote.market_price:.2f} "
                            f"({quote.change_pct:+.2f}%) — {quote.market_state}"
                        ),
                        model="rule-based-v1",
                    ),
                ))

        return events


# ── Brapi Crypto Ingestor ─────────────────────────────────


class BrapiCryptoIngestor(BaseIngestor):
    content_type = FinanceContentTypes.CRYPTO

    def __init__(self, signer: Any, coins: list[str] | None = None) -> None:
        super().__init__(signer)
        self.coins = coins or ["BTC", "ETH"]

    async def fetch(self) -> list[CDSEvent]:
        events: list[CDSEvent] = []
        coin_str = ",".join(self.coins)
        async with httpx.AsyncClient(timeout=15) as client:
            resp = await client.get(
                f"{BRAPI_BASE}/v2/crypto",
                params={"coin": coin_str, "currency": "BRL"},
            )
            resp.raise_for_status()
            fp = _fingerprint(resp.content)
            data = resp.json()

            for item in data.get("coins", []):
                payload = {
                    "coin": item.get("coin", ""),
                    "name": item.get("coinName", ""),
                    "currency": "BRL",
                    "price": float(item.get("regularMarketPrice", 0)),
                    "change_pct": float(item.get("regularMarketChangePercent", 0)),
                    "market_cap": item.get("marketCap"),
                    "volume_24h": item.get("regularMarketVolume"),
                    "timestamp": item.get("regularMarketTime", ""),
                }
                events.append(CDSEvent(
                    content_type=FinanceContentTypes.CRYPTO,
                    source=SourceMeta(id=CDSSources.BRAPI, fingerprint=fp),
                    occurred_at=datetime.now(UTC),
                    lang="pt-BR",
                    payload=payload,
                    event_context=ContextMeta(
                        summary=(
                            f"{payload['coin']}: R$ {payload['price']:,.2f} "
                            f"({payload['change_pct']:+.2f}%)"
                        ),
                        model="rule-based-v1",
                    ),
                ))

        return events


# ── Copom Ingestor ────────────────────────────────────────


class CopomIngestor(BaseIngestor):
    content_type = FinanceContentTypes.COPOM

    def __init__(self, signer: Any, last_n: int = 1) -> None:
        super().__init__(signer)
        self.last_n = last_n

    async def fetch(self) -> list[CDSEvent]:
        events: list[CDSEvent] = []
        async with httpx.AsyncClient(timeout=15) as client:
            resp = await client.get(COPOM_URL)
            resp.raise_for_status()
            fp = _fingerprint(resp.content)
            data = resp.json()

            meetings = data.get("conteudo", [])[:self.last_n]
            for item in meetings:
                rate_after = float(item.get("metaSelic", 0))
                rate_before = float(item.get("metaSelicAnterior", rate_after))
                diff_bps = round((rate_after - rate_before) * 100)

                if diff_bps > 0:
                    decision = "raise"
                elif diff_bps < 0:
                    decision = "cut"
                else:
                    decision = "maintain"

                meeting_date = item.get("dataReuniao", "")
                copom = CopomDecision(
                    meeting_number=int(item.get("numeroReuniao", 0)),
                    meeting_date=meeting_date,
                    decision=decision,
                    rate_before=rate_before,
                    rate_after=rate_after,
                    rate_change_bps=diff_bps,
                    vote_unanimous=item.get("unanime", True),
                    statement_url=item.get("urlAta"),
                )
                try:
                    occurred = datetime.fromisoformat(f"{meeting_date}T18:30:00-03:00")
                except (ValueError, TypeError):
                    occurred = datetime.now(UTC)

                events.append(CDSEvent(
                    content_type=FinanceContentTypes.COPOM,
                    source=SourceMeta(id=CDSSources.BCB_API, fingerprint=fp),
                    occurred_at=occurred,
                    lang="pt-BR",
                    payload=copom.model_dump(mode="json"),
                    event_context=ContextMeta(
                        summary=(
                            f"Copom #{copom.meeting_number}: "
                            f"{copom.decision} SELIC em {copom.rate_after}% a.a."
                        ),
                        model="rule-based-v1",
                    ),
                ))

        return events

File: sdk/typescript/src/sources/finance.ts
/**
 * SignedData CDS — Finance Brazil Source
 * Sources:
 *   BCB SGS API: https://api.bcb.gov.br/dados/serie/bcdata.sgs.{code}/dados/ultimos/{n}
 *   Brapi:       https://brapi.dev/api/quote/{tickers}
 */

import { createHash } from "node:crypto";
import { CDSEvent, ContextMeta, SourceMeta } from "../schema.js";
import { CDSSigner } from "../signer.js";
import { BaseIngestor } from "../ingestor.js";
import { CDSVocab, CDSSources } from "../vocab.js";

// ── Content Types ──────────────────────────────────────────

export const FinanceContentTypes = {
  SELIC:    CDSVocab.FINANCE_SELIC_RATE,
  CDI:      CDSVocab.FINANCE_CDI_RATE,
  IPCA:     CDSVocab.FINANCE_IPCA_INDEX,
  IGPM:     CDSVocab.FINANCE_IGPM_INDEX,
  USD_BRL:  CDSVocab.FINANCE_FX_USD_BRL,
  EUR_BRL:  CDSVocab.FINANCE_FX_EUR_BRL,
  STOCK:    CDSVocab.FINANCE_QUOTE_STOCK,
  FII:      CDSVocab.FINANCE_QUOTE_FII,
  CRYPTO:   CDSVocab.FINANCE_QUOTE_CRYPTO,
  COPOM:    CDSVocab.FINANCE_DECISION_COPOM,
} as const;

// ── Payload Types ──────────────────────────────────────────

export interface SELICRate {
  date: string;
  rate_annual: number;
  rate_daily: number;
  unit: string;
  effective_date: string;
}

export interface CDIRate {
  date: string;
  rate_annual: number;
  rate_daily: number;
  unit: string;
}

export interface IPCAIndex {
  date: string;
  monthly_pct: number;
  accumulated_12m: number;
  accumulated_year: number;
  base_year: number;
}

export interface IGPMIndex {
  date: string;
  monthly_pct: number;
}

export interface FXRate {
  date: string;
  buy: number;
  sell: number;
  mid: number;
  currency_from: string;
  currency_to: string;
  source: string;
}

export interface StockQuote {
  ticker: string;
  short_name: string;
  long_name: string;
  currency: string;
  market_price: number;
  change: number;
  change_pct: number;
  previous_close: number;
  open?: number;
  day_high: number;
  day_low: number;
  volume: number;
  market_cap?: number;
  exchange: string;
  market_state: string;
  timestamp: string;
}

export interface CopomDecision {
  meeting_number: number;
  meeting_date: string;
  decision: "raise" | "cut" | "maintain";
  rate_before: number;
  rate_after: number;
  rate_change_bps: number;
  vote_unanimous: boolean;
  statement_url?: string;
}

// ── Helpers ────────────────────────────────────────────────

const BCB_SGS_BASE = "https://api.bcb.gov.br/dados/serie/bcdata.sgs.{code}/dados/ultimos/{n}";
const BRAPI_BASE = "https://brapi.dev/api";

type R = Record<string, unknown>;

function bcbUrl(seriesCode: number, lastN: number = 1): string {
  return BCB_SGS_BASE.replace("{code}", String(seriesCode)).replace("{n}", String(lastN));
}

function parseBcbDate(raw: string): string {
  const [d, m, y] = raw.split("/");
  return d && m && y ? `${y}-${m}-${d}` : raw;
}

function dailyRate(annualPct: number): number {
  return ((1 + annualPct / 100) ** (1 / 252) - 1) * 100;
}

function fingerprint(data: Buffer): string {
  return "sha256:" + createHash("sha256").update(data).digest("hex");
}

// ── BCB Rates Ingestor ─────────────────────────────────────

export class BCBRatesIngestor extends BaseIngestor {
  readonly contentType = FinanceContentTypes.SELIC;
  private readonly lastN: number;

  constructor(signer: CDSSigner, lastN: number = 1) {
    super(signer);
    this.lastN = lastN;
  }

  async fetch(): Promise<CDSEvent[]> {
    const events: CDSEvent[] = [];

    // SELIC
    const selicResp = await globalThis.fetch(bcbUrl(11, this.lastN));
    if (!selicResp.ok) throw new Error(`BCB SGS HTTP ${selicResp.status}`);
    const selicBuf = Buffer.from(await selicResp.arrayBuffer());
    const selicFp = fingerprint(selicBuf);
    const selicData = JSON.parse(selicBuf.toString("utf-8")) as R[];

    for (const item of selicData) {
      const dateIso = parseBcbDate(item["data"] as string);
      const annual = Number(item["valor"]);
      const rate: SELICRate = {
        date: dateIso,
        rate_annual: annual,
        rate_daily: Math.round(dailyRate(annual) * 1e6) / 1e6,
        unit: "% a.a.",
        effective_date: dateIso,
      };
      events.push(new CDSEvent({
        content_type: FinanceContentTypes.SELIC,
        source: { "@id": CDSSources.BCB_API, fingerprint: selicFp },
        occurred_at: `${dateIso}T18:30:00-03:00`,
        lang: "pt-BR",
        payload: rate as unknown as R,
        context: {
          summary: `SELIC: ${rate.rate_annual}% a.a. (${rate.date})`,
          model: "rule-based-v1",
          generated_at: new Date().toISOString(),
        } satisfies ContextMeta,
      }));
    }

    // CDI
    const cdiResp = await globalThis.fetch(bcbUrl(4391, this.lastN));
    if (!cdiResp.ok) throw new Error(`BCB SGS HTTP ${cdiResp.status}`);
    const cdiBuf = Buffer.from(await cdiResp.arrayBuffer());
    const cdiFp = fingerprint(cdiBuf);
    const cdiData = JSON.parse(cdiBuf.toString("utf-8")) as R[];

    for (const item of cdiData) {
      const dateIso = parseBcbDate(item["data"] as string);
      const annual = Number(item["valor"]);
      const rate: CDIRate = {
        date: dateIso,
        rate_annual: annual,
        rate_daily: Math.round(dailyRate(annual) * 1e6) / 1e6,
        unit: "% a.a.",
      };
      events.push(new CDSEvent({
        content_type: FinanceContentTypes.CDI,
        source: { "@id": CDSSources.BCB_API, fingerprint: cdiFp },
        occurred_at: `${dateIso}T18:30:00-03:00`,
        lang: "pt-BR",
        payload: rate as unknown as R,
        context: {
          summary: `CDI: ${rate.rate_annual}% a.a. (${rate.date})`,
          model: "rule-based-v1",
          generated_at: new Date().toISOString(),
        } satisfies ContextMeta,
      }));
    }

    return events;
  }
}

// ── Brapi Quotes Ingestor ──────────────────────────────────

export interface BrapiQuotesOptions {
  tickers?: string[];
}

export class BrapiQuotesIngestor extends BaseIngestor {
  readonly contentType = FinanceContentTypes.STOCK;
  private readonly tickers: string[];

  constructor(signer: CDSSigner, opts: BrapiQuotesOptions = {}) {
    super(signer);
    this.tickers = opts.tickers ?? ["PETR4", "VALE3", "ITUB4"];
  }

  async fetch(): Promise<CDSEvent[]> {
    const events: CDSEvent[] = [];
    const tickerStr = this.tickers.join(",");

    const resp = await globalThis.fetch(`${BRAPI_BASE}/quote/${tickerStr}`);
    if (!resp.ok) throw new Error(`Brapi HTTP ${resp.status}`);
    const buf = Buffer.from(await resp.arrayBuffer());
    const fp = fingerprint(buf);
    const data = JSON.parse(buf.toString("utf-8")) as R;
    const results = (data["results"] as R[]) ?? [];

    for (const item of results) {
      const ticker = (item["symbol"] as string) ?? "";
      const quote: StockQuote = {
        ticker,
        short_name: (item["shortName"] as string) ?? "",
        long_name: (item["longName"] as string) ?? "",
        currency: (item["currency"] as string) ?? "BRL",
        market_price: Number(item["regularMarketPrice"] ?? 0),
        change: Number(item["regularMarketChange"] ?? 0),
        change_pct: Number(item["regularMarketChangePercent"] ?? 0),
        previous_close: Number(item["regularMarketPreviousClose"] ?? 0),
        open: item["regularMarketOpen"] as number | undefined,
        day_high: Number(item["regularMarketDayHigh"] ?? 0),
        day_low: Number(item["regularMarketDayLow"] ?? 0),
        volume: Number(item["regularMarketVolume"] ?? 0),
        market_cap: item["marketCap"] as number | undefined,
        exchange: (item["fullExchangeName"] as string) ?? "SAO",
        market_state: (item["marketState"] as string) ?? "CLOSED",
        timestamp: (item["regularMarketTime"] as string) ?? new Date().toISOString(),
      };

      const ct = ticker.endsWith("11") && ticker.length === 6
        ? FinanceContentTypes.FII
        : FinanceContentTypes.STOCK;

      let occurred: string;
      try {
        occurred = new Date(quote.timestamp).toISOString();
      } catch {
        occurred = new Date().toISOString();
      }

      events.push(new CDSEvent({
        content_type: ct,
        source: { "@id": CDSSources.BRAPI, fingerprint: fp },
        occurred_at: occurred,
        lang: "pt-BR",
        payload: quote as unknown as R,
        context: {
          summary: `${quote.ticker}: R$ ${quote.market_price.toFixed(2)} (${quote.change_pct >= 0 ? "+" : ""}${quote.change_pct.toFixed(2)}%) — ${quote.market_state}`,
          model: "rule-based-v1",
          generated_at: new Date().toISOString(),
        } satisfies ContextMeta,
      }));
    }

    return events;
  }
}
